triopg/_triopg.py

```diff
@@ -121,6 +122,27 @@ async def prepare(self, *args, **kwargs):
         )
         return TrioStatementProxy(asyncpg_statement)
 
+    @asynccontextmanager
+    async def listen(self, channel):
+        """LISTEN on `channel` notifications and return memory channel to iterate over
+
+        For example:
+
+        async with conn.listen('some.changes') as changes:
+            async for change in changes:
+                print('Postgres notification received:', change)
+        """
+
+        send_channel, receive_channel = trio.open_memory_channel(1)
+
+        def _listen_callback(c, pid, chan, payload):
+            send_channel.send_nowait(payload)
```

Comment:
Another problem I've missed before: `.send_nowait()` could raise `trio.WouldBlock`, that has to be handled somehow:

* I couldn't send it over the memory channel, because it's (obviously) full.
* I don't want to set memory channel size to `inf`, because it would leak memory if you forget to consume notifications.

The best option is to raise `TooSlowToConsumeNotification` error somewhere. But where should the error go?..

Reply:
I've asked for help in gitter: https://gitter.im/python-trio/general?at=5ffb5c3ed5f4bf2965da7232

Reply:
Good catch indeed !
The trouble I see here is the handling of this back pressure is really user dependent so there is no one-size-fits-all solution...

I guess adding a `queue_size` parameter (defaulting to `math.inf`) in the `listen()` method would be a good step toward solving the issue.

BTW I wonder what happens when the `_listen_callback` raises an exception, does asyncpg catches it (maybe displaying an error/warning log) or does it blows the entire connection altogether ?

In the end maybe the best solution is to acknowledge `listen` is just a helper covering with limited usecase (then typically setting the queue size to `inf` so we solve the exception issue) that should be well documented (and encourage user with other usecase to take inspiration of the code and build his own solution from it)

Reply:
Ok, I've fixed it with a `CancelScope`.

Reply:
Sorry, I haven't noticed your message before now. GitHub managed to hide it from me somehow 🤷‍♂️ 

I think we can provide good back pressure solution without much configurability. If we can ask tell the socket to stop flooding us with notifications, it should work for almost all cases. @njsmith argues for this point in his 2016 blog post: https://vorpus.org/blog/some-thoughts-on-asynchronous-api-design-in-a-post-asyncawait-world/

Unfortunately, it's unclear how to do it with `asyncpg`. See, for example, MagicStack/asyncpg#463. Fortunately, today's discussion on Trio gitter made me think it's not over.

Reply:
Regarding the exception in `_listen_callback`, I haven't seen anything in the logs or elsewhere. Maybe it's due `trio-asyncio` library limitations. It does have bugs.

Reply:
@dhirschfeld [has shared][1] his `LISTEN` code, using `psycopg2` library. It's very nice and it does support backpressure out of the box. No need to worry about weird buffer size, or the exceptions.

@touilleMan, it's very tempting to re-implement this `conn.listen` helper using `psycopg2` instead of `asyncpg`. Even though bringing another dependency is not ideal...

[1]: https://gitter.im/python-trio/general?at=5ffbf02a91e9b71badcb8ff5

Reply:
for me adding `psycopg2` in this lib for just a helper is a big no (this will open the pandora box of having the need to bring a full binding to psycopg2, and it will add a lot of complexity to handle psycopg2 connections the same way asyncpg connections are)

Reply:
@touilleMan Yes, of course :-) My comment was mostly to signal a frustration regarding limitations of different libraries. I think I'll implement @njsmith suggestion.